blockchain/blockdao/blob_store.go

```diff
@@ -22,4 +38,241 @@ type (
+	//    purging expired data.
+	// 2. Two mappings are kept in a separate index: (1) key to blob's height
+	//    and tx hashes, and (2) tx hash to the blob's height. Here's an example:
+	//    304 --> { height: 311344
+	//              []hash: eddf9e61fb9d8f5111840daef55e5fde
+	//                      95977fb340867fa13f787df864ed5388 }
+	//    and:
+	//    eddf9e61fb9d8f5111840daef55e5fde --> {height: 311344}
+	//    95977fb340867fa13f787df864ed5388 --> {height: 311344}
+	//    The 2nd mapping is used to support GetBlob(hash) function. It converts
+	//    the requested blob hash to the height, and retrieves the blob.
+	//    When a new blob overwrites an old one, it will check the 1st mapping,
+	//    and delete the hashes of old blob correspondingly.
+	// 3. The maximum number of blobs is 6 for each block, so maximum data size
+	//    stored by a key is 131kB x 6 = 786kB. The maximum total size of the
+	//    entire blob storage is 786kB x 311040 = 245GB.
+	//
+	blobStore struct {
+		kvStore                         db.KVStore
+		epochNumber, blockPerEpoch      uint32
+		totalBlocks                     uint64
+		currWriteBlock, currExpireBlock uint64
+	}
 )
+
+func NewBlobStore(kv db.KVStore, epochNum, blockPerEpoch uint32) *blobStore {
+	return &blobStore{
+		kvStore:       kv,
+		epochNumber:   epochNum,
+		blockPerEpoch: blockPerEpoch,
+		totalBlocks:   uint64(epochNum) * uint64(blockPerEpoch),
+	}
+}
+
+func (bs *blobStore) Start(ctx context.Context) error {
+	if err := bs.kvStore.Start(ctx); err != nil {
+		return err
+	}
+	return bs.checkDB()
+}
+
+func (bs *blobStore) Stop(ctx context.Context) error {
+	if err := bs.kvStore.Stop(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bs *blobStore) checkDB() error {
+	var err error
+	bs.currWriteBlock, err = bs.getHeightByHash(_writeHeight)
+	if err != nil && errors.Cause(err) != db.ErrNotExist {
+		return err
+	}
+	bs.currExpireBlock, err = bs.getHeightByHash(_expireHeight)
+	if err != nil && errors.Cause(err) != db.ErrNotExist {
+		return err
+	}
+	return nil
+}
+
+func (bs *blobStore) GetBlob(h hash.Hash256) (*types.BlobTxSidecar, string, error) {
+	height, err := bs.getHeightByHash(h[:])
+	if err != nil {
+		return nil, "", err
+	}
+	blobs, hashes, err := bs.getBlobs(height)
+	if err != nil {
+		return nil, "", err
+	}
+	for i, v := range hashes {
+		if common.BytesToHash(h[:]).Hex() == v {
+			return blobs[i], v, nil
+		}
+	}
+	return nil, "", errors.Wrapf(db.ErrNotExist, "blob hash = %x", h)
+}
+
+func (bs *blobStore) GetBlobsByHeight(height uint64) ([]*types.BlobTxSidecar, []string, error) {
+	return bs.getBlobs(height)
+}
+
+func (bs *blobStore) getBlobs(height uint64) ([]*types.BlobTxSidecar, []string, error) {
+	storeHeight, _, err := bs.getBlobIndex(height)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to get blob index at height %d", height)
+	}
+	if storeHeight != height {
+		// the stored item does not match the request height
```

Comment:
This could be a serious issue, and we need to be aware of it.

Reply:
no longer needed